### ISSUE TITLE
Crash on decoding WHEA_XPF_MCA_SECTION record

### ISSUE THREAD

**Issue by bootsareme** [ralish/DecodeWheaRecord#6]
WHEA-Logger displays these events as 
> A fatal hardware error has occurred. A record describing the condition is contained in the data section of this event.

I wrote a script in Python to parse through the raw data:

```py
def bytes2num(byte_str):
    return str(int.from_bytes(bytes.fromhex(byte_str), byteorder="little"))


raw_whea = "435045521002FFFFFFFF04000100000002000000FB03000014140E00180A18140000000000000000000000000000000000000000000000000000000000000000BDC407CF89B7184EB3C41F732CB57131FE6FF5E89C91C54CBA8865ABE14913BBC0D5DAD91F26DB01020000000000000000000000000000000000000000000000A00100005000000000030000010000001411BCA5646FDE4EB8633E83ED7C83B100000000000000000000000000000000010000000000000000000000000000000000000000000000F0010000C00000000003000000000000ADCC7698B447DB4BB65E16F193C4F3DB00000000000000000000000000000000010000000000000000000000000000000000000000000000B0020000240100000003000000000000011D1E8AF94257459C33565E5CC3F7E800000000000000000000000000000000010000000000000000000000000000000000000000000000D4030000270000000003000000000000A13248C3C302524CA9F19F1D5D7723FC0000000000000000000000000000000003000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007F010000000000000002010100010000100FA2000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000B00000000000000000000000000000000000000000000000000000000000000000000000000000003000000020000002C9E4CDB1F26DB010B000000000000000000000000000000000000000000000035010C000000A0BA000000000000000000000000FF0F13D00A0000000B00000000000000B00010001400004D00000000FD010000070000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001B00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000FF00000000000000000000000000000000000000000000000000"
print("Error severity: " + bytes2num(raw_whea[12*2:12*2+4*2]))
print("Record length: " + bytes2num(raw_whea[20*2:20*2+4*2]))
print("Notification type: " + little2big_endian(raw_whea[80*2:80*2+16*2]))
print("Flags: " + bytes2num(raw_whea[104*2:104*2+4*2]))

section_descriptor = raw_whea[128*2:]
print("Section count: " + bytes2num(raw_whea[10*2:10*2+2*2]))
print("Section type: " + section_descriptor[16*2:16*2+16*2])

section_body = raw_whea[128*2+72*2*4:]
print("Error status: " + section_body[8*2:8*2+8*2])
print("Physical address: " + section_body[16*2:16*2+8*2])
```
However, I gave up when I could not produce any useful information other than:
- Notification type: **Machine Check Exception** `{0xE8F56FFE, 0x919C, 0x4cc5, {0xBA, 0x88, 0x65, 0xAB, 0xE1, 0x49, 0x13, 0xBB}}`
- Section type: **Platform Memory** `{0xA5BC1114, 0x6F64, 0x4EDE, {0xB8, 0x63, 0x3E, 0x83, 0xED, 0x7C, 0x83, 0xB1}}`

I knew the above was true because I referenced Appendix N: Common Platform Error Record of https://uefi.org/sites/default/files/resources/UEFI_Spec_2_2_D.pdf. I now want to know more about this error. That is when I found your tool. I ran the tool with the raw data but it crashes with: 

```
[WHEA_ERROR_RECORD] Header indicates error record contains 1019 bytes but marshalled 999 bytes.

Unhandled Exception: Newtonsoft.Json.JsonSerializationException: Error getting value from 'AmdBits' on 'DecodeWheaRecord.NativeMethods+WHEA_XPF_MCA_SECTION'. ---> System.InvalidCastException: Unable to cast object of type 'MCI_STATUS_INTEL_BITS' to type 'MCI_STATUS_AMD_BITS'.
   at GetAmdBits(Object )
   at Newtonsoft.Json.Serialization.DynamicValueProvider.GetValue(Object target)
   --- End of inner exception stack trace ---
   at Newtonsoft.Json.Serialization.DynamicValueProvider.GetValue(Object target)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.CalculatePropertyValues(JsonWriter writer, Object value, JsonContainerContract contract, JsonProperty member, JsonProperty property, JsonContract& memberContract, Object& memberValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeObject(JsonWriter writer, Object value, JsonObjectContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeList(JsonWriter writer, IEnumerable values, JsonArrayContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeObject(JsonWriter writer, Object value, JsonObjectContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.Serialize(JsonWriter jsonWriter, Object value, Type objectType)
   at Newtonsoft.Json.JsonSerializer.SerializeInternal(JsonWriter jsonWriter, Object value, Type objectType)
   at Newtonsoft.Json.JsonConvert.SerializeObjectInternal(Object value, Type type, JsonSerializer jsonSerializer)
   at DecodeWheaRecord.Program.Main(String[] args)
```
Any insights you can provide? Does the record decoder not support AMD chips? That is what I deduced from the exception.

**Comment by ralish** [comment]
It's not so much due to a lack of AMD support but that I haven't fully implemented support for decoding that error type. It's one of the more complex errors types to decode and I didn't have any sample error records with which I could easily test against.

Using your record I've implemented near complete support. I'll push my changes once I've done a bit of clean-up to the code but here's what I now get for your WHEA record:

```
{
  "Header": {
    "Signature": "CPER",
    "Revision": "2.16",
    "SignatureEnd": "0xffffffff",
    "SectionCount": 4,
    "Severity": "Fatal",
    "ValidBits": "Timestamp",
    "Length": 1019,
    "Timestamp": "24/10/2024 2:20:20 PM",
    "CreatorId": "Microsoft",
    "NotifyType": "Machine Check Exception (MCE)",
    "RecordId": 133742532178400704,
    "Flags": "PreviousError",
    "Reserved": "AAAAAAAAAAA="
  },
  "SectionDescriptor": [
    {
      "SectionOffset": 416,
      "SectionLength": 80,
      "Revision": "3.0",
      "ValidBits": "",
      "Reserved": 0,
      "Flags": "Primary",
      "SectionType": "Memory Error",
      "SectionSeverity": "Fatal"
    },
    {
      "SectionOffset": 496,
      "SectionLength": 192,
      "Revision": "3.0",
      "ValidBits": "",
      "Reserved": 0,
      "Flags": "",
      "SectionType": "Generic Processor Error",
      "SectionSeverity": "Fatal"
    },
    {
      "SectionOffset": 688,
      "SectionLength": 292,
      "Revision": "3.0",
      "ValidBits": "",
      "Reserved": 0,
      "Flags": "",
      "SectionType": "IA32/AMD64 Machine Check Error",
      "SectionSeverity": "Fatal"
    },
    {
      "SectionOffset": 980,
      "SectionLength": 39,
      "Revision": "3.0",
      "ValidBits": "",
      "Reserved": 0,
      "Flags": "",
      "SectionType": "Error Recovery Information",
      "SectionSeverity": "Informational"
    }
  ],
  "Section": [
    {
      "ValidBits": ""
    },
    {
      "ValidBits": "ProcessorType, InstructionSet, ErrorType, Operation, Flags, Level, CPUVersion, ProcessorId",
      "ProcessorType": "XPF",
      "InstructionSet": "X64",
      "ErrorType": "Cache",
      "Operation": "DataRead",
      "Flags": "",
      "Level": 1,
      "CPUVersion": {
        "Stepping": 0,
        "Model": 1,
        "Family": 15,
        "ProcessorType": 0,
        "ExtendedModel": 2,
        "ExtendedFamily": 10
      },
      "ProcessorId": 11
    },
    {
      "VersionNumber": 3,
      "CpuVendor": "Amd",
      "Timestamp": 133742532202634796,
      "ProcessorNumber": 11,
      "GlobalStatus": "",
      "InstructionPointer": 0,
      "BankNumber": 0,
      "AmdBits": {
        "McaErrorCode": 309,
        "ModelErrorCode": 12,
        "Flags": "ContextCorrupt, MiscValid, ErrorEnabled, UncorrectedError, Valid",
        "ImplementationSpecific1": 2560,
        "ImplementationSpecific2": 0
      },
      "Address": 0,
      "Misc": "0xd0130fff00000000",
      "ExtendedRegisterCount": 10,
      "ApicId": 11,
      "AMDExtendedRegisters": {
        "IPID": "0x1000b000000000",
        "SYND": "0x4d000014",
        "CONFIG": "0x7000001fd",
        "DESTAT": 0,
        "DEADDR": 0,
        "MISC1": 0,
        "MISC2": 0,
        "MISC3": 0,
        "MISC4": 0,
        "RasCap": 27
      },
      "GlobalCapability": {
        "Flags": "",
        "CountField": 0,
        "ExtendedRegisterCount": 0
      },
      "RecoveryInfo": {
        "FailureReason": "",
        "Action": "",
        "ActionRequired": false,
        "RecoverySucceeded": false,
        "RecoveryKernel": false
      }
    },
    {
      "RecoveryKernel": false,
      "RecoveryAction": "NoneAttempted",
      "RecoveryType": null,
      "Irql": 255,
      "RecoverySucceeded": false,
      "FailureReason": null,
      "ProcessName": ""
    }
  ]
}
```

As for actually understanding the content I'll leave that to you. It's clearly a CPU exception and the most relevant part is probably:

```
"AmdBits": {
  "McaErrorCode": 309,
  "ModelErrorCode": 12,
  "Flags": "ContextCorrupt, MiscValid, ErrorEnabled, UncorrectedError, Valid",
  "ImplementationSpecific1": 2560,
  "ImplementationSpecific2": 0
},
```

You'd need to consult the AMD technical documentation for actually finding out what those error codes mean.

**Comment by bootsareme** [comment]
@ralish Thank you for decoding the error for me. Is the fix pushed onto the main branch? I also believe you should generate a release binary so that users will not have to clone and build from source.

**Comment by ralish** [comment]
Yes, the code is now on `main` and I've published a release as v0.3.0. Note this project is very much still in an alpha quality state so crashes and other breakage are to be expected.
